packages/amesos2/src/Amesos2_Tpetra_Impl.hpp

```diff
@@ -0,0 +1,376 @@
+// @HEADER
+//
+// ***********************************************************************
+//
+//           Amesos2: Templated Direct Sparse Solver Package
+//                  Copyright 2011 Sandia Corporation
+//
+// Under the terms of Contract DE-AC04-94AL85000 with Sandia Corporation,
+// the U.S. Government retains certain rights in this software.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+// 1. Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY SANDIA CORPORATION "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL SANDIA CORPORATION OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Michael A. Heroux (maherou@sandia.gov)
+//
+// ***********************************************************************
+//
+// @HEADER
+
+/**
+  \file   Amesos2_Tpetra_Impl.hpp
+  \author
+  \date
+
+  \brief  ETI for Solvers using Tpetra adapter
+*/
+
+#ifndef AMESOS2_TPETRA_IMPL_HPP
+#define AMESOS2_TPETRA_IMPL_HPP
+
+// MDM-DISCUSS Moved this Tpetra ETI into a dedicated file so we could share it
+// for different solvers. Decide if this is ok. As we add the other solvers we
+// can switch them to this file to avoid the duplication.
+
+#ifdef HAVE_TPETRA_INST_INT_INT
+#ifdef HAVE_TPETRA_INST_FLOAT
+  AMESOS2_SOLVER_TPETRA_INST(AMESOS2_TPETRA_IMPL_SOLVER_NAME,float,int,int);
```

Comment:
Doesn't Amesos2 have a macro now for doing instantiations?

Reply:
@kyungjoo-kim I see now superlu was using TPETRA_INSTANTIATE_SLGN_NO_ORDINAL_SCALAR but the other solvers manually implement the ETI. I changed Tacho to also use TPETRA_INSTANTIATE_SLGN_NO_ORDINAL_SCALAR. Then I can remove the Amesos2_Tpetra_Impl.hpp file. If we want all solvers like this I could migrate them as I do the follow up PRs. If this is a problem I will just roll back.  @mhoemmen Not sure if this is what you meant. I don't know if there are any higher level macros I should be using in Amesos2.